### DISCUSSION TITLE
Store and read light sensor values with Spike Prime - system.storage

### DISCUSSION THREAD

**Discussion by Aldeatron** [pybricks/support#1167] [General]
Hi everyone!

We try to store the B&W light intensity values and read later in a line follower code with "system.storage". We don´t quite understand how to store and read binary data and then convert it to a number we could use as light intensity. Can anyone explain how it can be done and give an example if possible? 

Thank you!

**Comment by laurensvalk** [comment]
For small numbers (between 0 and 255), not much conversion is required. Here is an example:

```python
from pybricks.hubs import PrimeHub

hub = PrimeHub()

# In this example we alway store it at the start of the memory.
START = 0

# Each number may be between 0 and 255
write_values = [7, 8, 9, 35, 200]
print("I store", write_values)
hub.system.storage(START, write=bytes(write_values))

# You can read it now. Even if you delete the values above
# the second time. Even after restart.
read_values = [v for v in  hub.system.storage(START, read=5)]
print("I read ", read_values)

```

**Reply by Aldeatron** [reply]
Thank you very much!

**Comment by Aldeatron** [comment]
Hi! How could we store and read float numbers?

**Reply by laurensvalk** [reply]
You can convert them to bytes:
```python
import ustruct

# This is the format for 3 floating point values
# See documentation for other data types.
FORMAT = "fff"

# Convert to bytes.
packed_data = ustruct.pack(FORMAT, 1.23, 3.14, -5.6e8)

# You can store/load them to memory now.
print("bytes", packed_data)

# This is how you get the values back.
unpacked_data = ustruct.unpack(FORMAT, packed_data)
print("unpacked again", unpacked_data)

```

Or another way to look at the problem: Do you really need to store float numbers, or will small int values be enough ?

Earlier you mentioned the need to scale sensor values, so you wanted to store the value `2.5` in order to scale 40 to 100.  Another way to think about this particular problem is just to store the value `40` so you don't have to do any conversions.

**Reply by Aldeatron** [reply]
Thank you! 

> Or another way to look at the problem: Do you really need to store float numbers, or will small int values be enough ?

That´s what we did. We stored factors as integers (2.5 -> 25) and then converted them again in the program we use the sensors (i.e, line follower). But we want to try with more precision. It´s not a bad idea to store the readings, but then we would have to calculate the same factors in every program we use the sensors.

**Reply by Aldeatron** [reply]
Hi! Could you please tell us how to unpack values in another program? 

Thank you!

**Reply by dlech** [reply]
The unpack in the example above will still work in other programs (as long as it matches the last pack used)

**Reply by Aldeatron** [reply]
Ok, thanks. But the new program doesn´t recognize "packed_data". How can we define it in the new program?

**Reply by dlech** [reply]
```python
packed_data = hub.system.storage(START, read=ustruct.calcsize(FORMAT))
```

**Reply by Aldeatron** [reply]
Thanks!

**Reply by Aldeatron** [reply]
Hi again! Sorry about so many questions. 

This is the way we store our 4 float values:

![imagen](https://github.com/pybricks/support/assets/139243532/03be8916-726d-4093-88f3-d5903259a2c3)

We have tried this so far to unpack them:

![imagen](https://github.com/pybricks/support/assets/139243532/1e6af7c6-8992-4cd7-bd40-04eceffe3379)

But when we print them we obtain the integer values stored in previous tests, not the float ones.

Thank you!

**Reply by laurensvalk** [reply]
I don't seem to see a `write` command anywhere, e.g.:

```python
hub.system.storage(START, write=bytes(write_values))
```

That said, if you are essentially storing information about light sensor values (which can only measure from 0--100), I'll probably still recommend just storing those :)

**Reply by Aldeatron** [reply]
Thank you! 

We also need to store correction factors (CorrecE & CorrecF), which are float numbers.

**Reply by dlech** [reply]
FYI, you can encode more types (int, str) using `ustruct.pack()`. See the table at the bottom of https://docs.pybricks.com/en/stable/micropython/ustruct.html.
